bot/cogs/Calculator.py

```diff
@@ -0,0 +1,153 @@
+
+        op1 = self.search_operators_symbol(operator1)
+
+        op2 = self.search_operators_symbol(operator2)
+
+        return op1['precedence'] <= op2['precedence']
+
+    # searches through a list operators and return its information
+    def search_operators_symbol(self, symbol):
+        for operator in self.operators:
+            if symbol == operator['symbol']:
+                return operator
+
+    # checks if symbol is an operator
+    def is_operator(self, symbol):
+        for op in self.operators:
+            if symbol == op['symbol']:
+                return True
+        return False
+
+    # returns the last element in a list
+    def get_top_stack(self, stack):
+        return stack[len(stack) - 1]
+
+    def parse_expression(self, expression):
+
+        # parse expression into a list of numbers and symbols
+        tokens = re.findall("-?\d+|[+^/*()-]", expression)
+
+        output_que = []
+
+        operator_stack = []
+
+        for token in tokens:
+
+            intFlag = 1
+            try:
+                int(token)
+            except ValueError:
+                intFlag = 0
+
+            if intFlag:
+                output_que.append(token)
+
+            elif token == '(':
+                operator_stack.append(token)
+
+            elif token == ')':
+                # Ensures that proper precedence is followed with parentheses
+                while self.get_top_stack(operator_stack) != '(':
+                    output_que.append(self.get_top_stack(operator_stack))
+                    operator_stack.pop()
+                operator_stack.pop()
+
+            elif self.is_operator(token):
+                # Makes sure precedence is followed for operators
+                while (len(operator_stack) != 0 and (self.get_top_stack(operator_stack) not in "()") 
+                and self.compare_precedence(token, self.get_top_stack(operator_stack)) 
+                and self.search_operators_symbol(token)['assoc'] == 'L'):
+                    output_que.append(self.get_top_stack(operator_stack))
+                    operator_stack.pop()
+                operator_stack.append(token)
+
+        while len(operator_stack) != 0:
+            output_que.append(self.get_top_stack(operator_stack))
+            operator_stack.pop()
+
+        return output_que
+
+    def math(self, num1, num2, sign):
+
+        if sign == "+":
+            result = num1 + num2
+        elif sign == "-":
+            result = num1 - num2
+        elif sign == "*":
+            result = num1 * num2
+        elif sign == '^':
+            result = num1 ** num2
+        else:
+            result = num1 / num2
+
+        return result
+
+    def parse_postfix(self, expression):
+
+        # stores numbers to calculated
+        num_stack = []
+
+        result = 0
+
+        for token in expression:
+
+            intFlag = 1
+            try:
+                int(token)
+            except ValueError:
```

Comment:
This sort of input validation should be done ideally at the top level so that you can respond to the user with why their command was invalid, relying on the exception bubbling to the top isnt reliable. responding with an accurate reason for failure is important. and should be done BEFORE the actual calculate method is called.

Reply:
That try and except is not meant for input validation but to check if the token passed is an int or not. I would use str.isnumeric() but that does not support negative numbers.